crates/ide/src/goto_implementation.rs

```diff
@@ -21,55 +24,55 @@ pub(crate) fn goto_implementation(
     let source_file = sema.parse(position.file_id);
     let syntax = source_file.syntax().clone();
 
-    let krate = sema.to_module_def(position.file_id)?.krate();
-
-    if let Some(nominal_def) = find_node_at_offset::<ast::Adt>(&syntax, position.offset) {
-        return Some(RangeInfo::new(
-            nominal_def.syntax().text_range(),
-            impls_for_def(&sema, &nominal_def, krate)?,
-        ));
-    } else if let Some(trait_def) = find_node_at_offset::<ast::Trait>(&syntax, position.offset) {
-        return Some(RangeInfo::new(
-            trait_def.syntax().text_range(),
-            impls_for_trait(&sema, &trait_def, krate)?,
-        ));
-    }
+    let module = sema.to_module_def(position.file_id)?;
+    let krate = module.krate();
 
-    None
+    let node = sema.find_node_at_offset_with_descend(&syntax, position.offset)?;
+    let def = match &node {
+        ast::NameLike::Name(name) => {
+            NameClass::classify(&sema, name).map(|class| class.referenced_or_defined(sema.db))
+        }
+        ast::NameLike::NameRef(name_ref) => {
+            NameRefClass::classify(&sema, name_ref).map(|class| class.referenced(sema.db))
+        }
+        ast::NameLike::Lifetime(_) => None,
+    }?;
+    let def = match def {
+        Definition::ModuleDef(def) => def,
+        _ => return None,
+    };
+    let navs = match def {
+        hir::ModuleDef::Trait(trait_) => impls_for_trait(&sema, trait_, krate),
+        hir::ModuleDef::Adt(adt) => impls_for_ty(&sema, adt.ty(sema.db), krate),
+        hir::ModuleDef::TypeAlias(alias) => impls_for_ty(&sema, alias.ty(sema.db), krate),
+        hir::ModuleDef::BuiltinType(builtin) => {
+            impls_for_ty(&sema, builtin.ty(sema.db, module), krate)
+        }
+        _ => return None,
+    };
+    Some(RangeInfo { range: node.syntax().text_range(), info: navs })
 }
 
-fn impls_for_def(
+fn impls_for_ty(
     sema: &Semantics<RootDatabase>,
-    node: &ast::Adt,
+    ty: hir::Type,
     krate: Crate,
-) -> Option<Vec<NavigationTarget>> {
-    let ty = match node {
-        ast::Adt::Struct(def) => sema.to_def(def)?.ty(sema.db),
-        ast::Adt::Enum(def) => sema.to_def(def)?.ty(sema.db),
-        ast::Adt::Union(def) => sema.to_def(def)?.ty(sema.db),
-    };
-
+) -> Vec<NavigationTarget> {
     let impls = Impl::all_in_crate(sema.db, krate);
-
-    Some(
-        impls
-            .into_iter()
-            .filter(|impl_def| ty.is_equal_for_find_impls(&impl_def.target_ty(sema.db)))
-            .filter_map(|imp| imp.try_to_nav(sema.db))
-            .collect(),
-    )
+    impls
+        .into_iter()
+        .filter(|impl_def| ty.is_equal_for_find_impls(&impl_def.target_ty(sema.db)))
+        .filter_map(|imp| imp.try_to_nav(sema.db))
+        .collect()
```

Comment:
This is currently wrong, as this only checks the inherent impls of the crate where the identifier lies in right now. Not of the definition, trying to figure out how to solve this for the builtin case as we don't have a defining crate

Reply:
`hir::Type` has a krate field which isnt publicly accessible at the moment, giving that an accessor and using that would fix the general problem, thought the builtin would still not work properly as we create it from the current modules resolver still. So I guess for this we need to fetch the krate id of core somehow?

Reply:
The `krate` on `hir::Type` is not necessarily the crate in which the struct/enum is defined, if the type is one. It's the crate where the type comes from (i.e. if you have a `let v: Vec<_>` in your code, the krate will be your crate).

Reply:
Ah okay, ye that makes sense given BuiltinTypes dont have a crate. Hmm

Reply:
See `Ty::def_crates` to actually get the crates where a type can have inherent impls.